Google.Api.Generator/Generation/MethodDetails.cs

```diff
@@ -232,13 +241,79 @@ private MethodDetails(ServiceDetails svc, MethodDescriptor desc)
+            desc.CustomOptions.TryGetMessage<HttpRule>(ProtoConsts.MethodOption.HttpRule, out var http);
             // Assume HTTP GET methods are idempotent; all others are non-idempotent.
-            IsIdempotent = desc.CustomOptions.TryGetMessage<HttpRule>(
-                ProtoConsts.MethodOption.HttpRule, out var http) ? !string.IsNullOrEmpty(http.Get) : false;
+            IsIdempotent = http != null ? !string.IsNullOrEmpty(http.Get) : false;
             DocLines = desc.Declaration.DocLines().ToList();
             Signatures = desc.CustomOptions.TryGetRepeatedMessage<MethodSignature>(ProtoConsts.MethodOption.MethodSignature, out var sigs) ?
                 sigs.Select(sig => new Signature(svc, desc.InputType, sig)).ToList() : (IReadOnlyList<Signature>)new Signature[0];
             RequestMessageDesc = desc.InputType;
+            RoutingHeaders = ReadRoutingHeaders(http, desc.InputType).ToList();
+        }
+
+        private IEnumerable<RoutingHeader> ReadRoutingHeaders(HttpRule http, MessageDescriptor requestDesc)
+        {
+            if (http != null)
+            {
+                // Read routing headers(s) from any of the http urls
+                var url =
+                    !string.IsNullOrEmpty(http.Get) ? http.Get :
+                    !string.IsNullOrEmpty(http.Post) ? http.Post :
+                    !string.IsNullOrEmpty(http.Put) ? http.Put :
+                    !string.IsNullOrEmpty(http.Patch) ? http.Patch :
+                    !string.IsNullOrEmpty(http.Delete) ? http.Delete : null;
+                if (url != null)
+                {
+                    foreach (var path in ExtractBracedPaths(url))
+                    {
+                        var desc = requestDesc;
+                        var propertyNames = new List<string>();
+                        FieldDescriptor finalField = null;
+                        foreach (var fieldName in path.Split('.'))
+                        {
+                            var field = finalField = desc.FindFieldByName(fieldName);
+                            desc = field.FieldType == FieldType.Message ? field.MessageType : null;
+                            if (field == null)
+                            {
+                                throw new InvalidOperationException($"Invalid path in http url: '{path}'. '{fieldName}' does not exist.");
+                            }
+                            propertyNames.Add(field.CSharpPropertyName());
+                        }
+                        if (finalField.FieldType != FieldType.String)
+                        {
+                            throw new InvalidOperationException($"Path in http url must resolve to a string field: '{path}'.");
+                        }
+                        yield return new RoutingHeader(WebUtility.UrlEncode(path), propertyNames);
+                    }
+                }
+            }
+
+            IEnumerable<string> ExtractBracedPaths(string s)
```

Comment:
I think you can use https://github.com/googleapis/gax-dotnet/blob/master/Google.Api.Gax/PathTemplate.cs instead.

Reply:
`PathTemplate` doesn't support the `{name=some.thing}` format. Trying to use it throws here: https://github.com/googleapis/gax-dotnet/blob/7dac7db0502d55eae3e9778d22d08e6112f5639b/Google.Api.Gax/PathTemplate.cs#L435

Reply:
Ah, interesting. Maybe we should fix that then. (Not that we use PathTemplate much, as far as I'm aware. But it is *meant* to support that annotation...)